Sources/NIOCore/Codec.swift

```diff
@@ -811,6 +818,9 @@ extension MessageToByteHandler {
             self.buffer!.clear()
             try self.encoder.encode(data: data, out: &self.buffer!)
             context.write(self.wrapOutboundOut(self.buffer!), promise: promise)
+            if let maxBufferCapacity {
```

Comment:
I am still a bit worried of doing this on every single write. Instead of adding some heuristic to only do It after a period of small writes.

Reply:
Easiest heuristic is to do it on `flush`.

Reply:
I have pushed a change, can you please take a look?

Reply:
Sorry, @FranzBusch  and @Lukasa wondering if I have hit the mark by clamping the buffer by using 

```
public func flush(context: ChannelHandlerContext) {
        context.flush()
        if let desiredBufferCapacity {
            self.buffer?.clampBufferCapacity(to: desiredBufferCapacity)
        }
    }
````

Let me know what you think when you find some time.